engine/apps/alerts/models/alert_group.py

```diff
@@ -141,7 +142,7 @@ def get_or_create_grouping(self, channel, channel_filter, group_data, received_a
                 web_title_cache=group_data.web_title_cache,
                 received_at=received_at,
             )
-            alert_group_created_signal.send(sender=self.__class__, alert_group=alert_group)
```

Comment:
this signal should be wrapped in a `transaction.on_commit` (to avoid having [this task](https://github.com/grafana/oncall/blob/dev/engine/apps/alerts/models/alert_group.py#L613) being occasionally unable to find the alert group object in the db) **but** if you look at the docstring for this method, it looks like we want to avoid wrapping the `create` call in a transaction?
> Also, "create" is invoked without transaction.atomic to reduce number of ConcurrentUpdateError's which can be
>         raised in AlertGroupQuerySet.create() due to optimistic locking of AlertGroupCounter model.

instead I think the `post_save` signal is probably our best bet here.

Reply:
> if you look at the docstring for this method

the comment says about the transaction inside the method `create`. This method is called before the signal, and no matter if it has transaction or not, signal will not be emmited if `create` failed.

This comment refers to the `transaction with select_for_update`, which could lock the table. Instead "optimistic locking" is used, check this file
https://github.com/grafana/oncall/blob/dev/engine/apps/alerts/models/alert_group_counter.py#L14

I'd prefer having the alert_group_created_signal.send called explicitly. We can move this signal to https://github.com/grafana/oncall/blob/e053eb084d84419159c0376be4de3a11a867e5ab/engine/apps/alerts/models/alert.py#L112 and wrap in `transaction.on_commit`, as we do later in the code

Reply:
what about this?:
```diff
-- alert_group_created_signal.send(sender=self.__class__, alert_group=alert_group)
++ transaction.on_commit(partial(alert_group_created_signal.send, sender=self.__class__, alert_group=alert_group))
```

I would argue it'd be beneficial to keep the `signal.send` call within the `get_or_create_grouping` method, because even though right now we're only calling `AlertGroupQuerySet.get_or_create_grouping` (within `Alert.create`), if in the future we call this method elsewhere, we'd have to remember to also call `signal.send` alongside the new function invocation (not sure if that makes sense?).

To be fair, I don't quite fully grok Django's `transaction.on_commit`.. In `Alert.create` we're not explicitly opening a transaction, so how does it know to open/close one? The [docs](https://docs.djangoproject.com/en/5.0/topics/db/transactions/#performing-actions-after-commit) say:
> If you call `on_commit()` while there isn’t an open transaction, the callback will be executed immediately.

Reply:
> To be fair, I don't quite fully grok Django's transaction.on_commit.. In Alert.create we're not explicitly opening a transaction, so how does it know to open/close one? The docs say

Right, there is no transaction yet, so the callback will be called immediately (as there is autocommit). We discussed similar problem in this https://github.com/grafana/oncall/pull/3731#issuecomment-1910899206

I want to add transaction (https://github.com/grafana/oncall/pull/3759/files#r1467196264) once verify it won't break `ConcurrentUpdateError` optimistic locking. 

I found it useful to check if there is an open transaction with
```
from django.db import transaction
print(transaction.get_connection().in_atomic_block)
```

Reply:
> I would argue it'd be beneficial to keep the signal.send call within the get_or_create_grouping method

That make sense, from the other hand I don't like this pattern as it is really easy to subscibe to that signal and unintentionally slow down alert processing